crates/autopilot/src/run_loop.rs

```diff
@@ -523,10 +537,39 @@ impl RunLoop {
         solutions
     }
 
+    /// Chooses the winners from the given participants.
+    ///
+    /// Participants are already sorted by their score (best to worst).
+    ///
+    /// Winners are selected one by one, starting from the best solution,
+    /// until `max_winners_per_auction` is hit. The solution can become winner
+    /// if it swaps tokens that are not yet swapped by any other already
+    /// selected winner.
+    fn select_winners<'a>(&self, participants: &'a VecDeque<Participant>) -> Vec<&'a Participant> {
```

Comment:
There is a significant difference that shadow competition works with `Result<Solution, Error>` and main competition works with `Solution`.

Reply:
Couldn't we just make `shadow` also work with `Solution` and not `Result` then?
Other than that detail it looks pretty similar, no?

Reply:
Pretty similar true, however different enough to have a different implementation with different structs for Participants etc... Will try to refactor everything but the PR will explode to > 500 changed lines...

Reply:
Tried to refactor and extract common code for main competition and shadow competition, but couldn't untangle properly. Will have to tackle this in a series of PRs.

Reply:
It would indeed be slightly more work than anticipated but let's make sure to have these follow up PRs as we should make it easy to keep the shadow and regular autopilot in sync. Also having these shared code paths should help us be more diligent about the core aspects of the autopilot which make up the most fundamental parts of the protocol.

Things that should be the same:
1. converting responses from the `/solve` API
2. discarding solutions based on errors (including deny-listed solvers)
3. choosing the winner
4. choosing the reference score

Things that should be different:
1. getting the auction
2. the whole `/settle` stuff
3. metrics and logging

Reply:
Yes, I detected this same list when tried to refactor the code. Will try to extract it and unify the competition for both modes in a follow ups